File: Ruby/lib/time_from_seconds.rb
# Class to convert seconds to days, hours and minutes
class TimeFromSeconds
  def initialize(seconds)
    @seconds = seconds
    @days = from_seconds_to_days
    @hours = from_seconds_to_hours
    @minutes = from_seconds_to_minutes
  end

  def to_puts
    to_puts = ''
    if @days.positive?
      to_puts << "#{@days} day#{@days > 1 ? 's' : ''}#{@hours.positive? || @minutes.positive? ? ', ' : ''}"
    end
    to_puts << if @hours.positive?
                 "#{@hours}:#{@minutes < 10 ? '0' : ''}#{@minutes}"
               else
                 "#{@minutes} min"
               end
    to_puts
  end

  private

  DAY = 86_400
  HOUR = 3600
  MINUTE = 60

  def from_seconds_to_days
    (@seconds / DAY).floor
  end

  def from_seconds_to_hours
    ((@seconds % DAY) / HOUR).floor
  end

  def from_seconds_to_minutes
    ((@seconds % DAY) % HOUR / MINUTE).floor
  end
end

File: Cpp/src/TimeFromSeconds.cpp
#include "../include/TimeFromSeconds.hpp"

TimeFromSeconds::TimeFromSeconds(int seconds) {
    this->seconds = seconds;
    this->days = this->fromSecondsToDays();
    this->hours = this->fromSecondsToHours();
    this->minutes = this->fromSecondsToMinutes();
}

std::string TimeFromSeconds::toCout() {
    std::string toReturn = "";
    if (this->days > 0) {
        toReturn.append(std::to_string(this->days) + " day");
        toReturn.append(this->days > 1 ? "s" : "");
        toReturn.append(this->hours > 0 || this->minutes > 0 ? ", " : "");
    }
    if (this->hours > 0) {
        toReturn.append(std::to_string(this->hours) + ":" + (this->minutes < 10 ? "0" : ""));
        toReturn.append(std::to_string(this->minutes));
    } else if (this->minutes > 0) {
        toReturn.append(std::to_string(this->minutes) + " min");
    }
    return toReturn;
}

const int TimeFromSeconds::DAY = 86400;
const int TimeFromSeconds::HOUR = 3600;
const int TimeFromSeconds::MINUTE = 60;

int TimeFromSeconds::fromSecondsToDays() {
    return this->seconds / TimeFromSeconds::DAY;
}

int TimeFromSeconds::fromSecondsToHours() {
    return (this->seconds % TimeFromSeconds::DAY) / TimeFromSeconds::HOUR;
}

int TimeFromSeconds::fromSecondsToMinutes() {
    return ((this->seconds % TimeFromSeconds::DAY) % TimeFromSeconds::HOUR) / TimeFromSeconds::MINUTE;
}
